### PR TITLE
feat(vanilla/react): New Jotai v2 API

### PR DESCRIPTION
**Pull request by dai-shi**
See the RFC: #1514 

- [x] initial implementation
- [x] convert tests/types.test.tsx
- [x] convert tests/basic.test.tsx
- [x] convert tests/async.test.tsx
- [x] convert tests/transition.test.tsx
- [x] convert tests/useAtomValue.test.tsx
- [x] convert tests/useSetAtom.test.tsx
- [x] convert tests/provider.test.tsx
- [x] convert tests/dependency.test.tsx
- [x] convert tests/items.test.tsx
- [x] convert tests/error.test.tsx
- [x] convert tests/onmount.test.tsx
- [x] convert tests/optimization.test.tsx
- [x] convert devtools
- [x] convert utils
- [x] migration guide `docs/guides/migrate-to-v2-api`
- [x] test with jotai-labs/jotai-location https://github.com/jotai-labs/jotai-location/pull/1
- [x] test with jotai-labs/jotai-form https://github.com/jotai-labs/jotai-form/pull/12
- [x] test with jotai-labs/jotai-tanstack-query https://github.com/jotai-labs/jotai-tanstack-query/pull/18
- [x] test with jotai-labs/jotai-urql https://github.com/jotai-labs/jotai-urql/pull/5
- [x] test with jotai-labs/jotai-relay https://github.com/jotai-labs/jotai-relay/pull/3
- [x] test wtih jotai-labs/jotai-optics https://github.com/jotai-labs/jotai-optics/pull/4
- [x] test with jotai-labs/jotai-immer https://github.com/jotai-labs/jotai-immer/pull/1
- [x] test with jotai-labs/jotai-trpc https://github.com/jotai-labs/jotai-trpc/pull/8
- [x] test with jotai-labs/jotai-xstate https://github.com/jotai-labs/jotai-xstate/pull/1
- [x] test with jotai-labs/jotai-redux https://github.com/jotai-labs/jotai-redux/pull/1
- [x] test with jotai-labs/jotai-zustand https://github.com/jotai-labs/jotai-zustand/pull/2
- [x] test with jotai-labs/jotai-valtio https://github.com/jotai-labs/jotai-valtio/pull/1
- [x] test with jotai-labs/jotai-jsx https://github.com/jotai-labs/jotai-jsx/pull/5
- [x] test with jotai-labs/jotai-game https://github.com/jotai-labs/jotai-game/pull/1
- [x] test with jotai-labs/jotai-suspense https://github.com/jotai-labs/jotai-suspense/pull/2
- [x] test with jotai-labs/jotai-cache https://github.com/jotai-labs/jotai-cache/pull/1

### CODE REVIEW

**Comment by dai-shi** [comment]
> With this change, would you recommend using Jotai over Valtio for React?

General suggestion won't change, but some users chose Zustand/Valtio over Jotai, only because Jotai states are not accessible outside React. For such users,  `createStore` API makes Jotai to be recommended.

File: docs/guides/migrating-to-async-api.mdx [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+Unfortunately, there are some breaking changes along with new features.
+
+## What are new features
+
+### Vanilla library
+
+Jotai comes with vanilla (non-React) functions
+and React functions separately.
+
+### Store I/F
```
Comment (Thisen) [inline docs/guides/migrating-to-async-api.mdx:22]: What is I/F?

File: docs/guides/migrating-to-async-api.mdx [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+  const store = useContext(MyContext)
+  useAtom(..., { store })
+```
+
+### `abortableAtom` util
+
+#### Previous API
+
+```js
+const asyncAtom = abortableAtom(async (get, { signal }) => {
```
Comment (Thisen) [inline docs/guides/migrating-to-async-api.mdx:160]: Maybe you should describe that `abortableAtom` is not needed, as async atoms has the `AbortController` built-in.

File: docs/guides/migrating-to-async-api.mdx [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+  ...
+}
+```
+
+### `waitForAll` util
+
+#### Previous API
+
+```js
+const allAtom = waitForAll([fooAtom, barAtom])
```
Comment (Thisen) [inline docs/guides/migrating-to-async-api.mdx:178]: Maybe you should describe that waitForAll is not needed, as it can be handled by native Promise APIs.

**Pull request merged**